superflore/test_integration/gentoo/build_base.py

```diff
@@ -26,18 +26,35 @@ def __init__(
         self.container = Docker()
         self.container.pull(image_owner, image_name)
         self.package_list = dict()
+        self.default_upstream = True
 
     def add_target(self, ros_distro, pkg):
         # TODO(allenh1): it might be nice to add a Python3 target
         # in case we want to test both.
         self.package_list['ros-%s/%s' % (ros_distro, pkg)] = 'unknown'
 
+    def set_upstream(self, upstream_url, branch_name='master'):
+        # check out the requested repo for testing
+        self.container.add_bash_command('cd /usr/local/portage')
+        self.container.add_bash_command(
+            'git remote add upstream %s' % upstream_url
+        )
+        self.container.add_bash_command('git fetch upstream')
+        self.container.add_bash_command(
+            'git checkout upstream/%s' % branch_name
+        )
+        self.container.add_bash_command(
+            'git reset --hard upstream/%s' % branch_name
+        )
+        self.default_upstream = False
+
     def run(self, verbose=True, log_file=None):
         # TODO(allenh1): add the ability to check out a non-master
         # branch of the overlay (for CI).
         info('testing gentoo package integrity')
         for pkg in sorted(self.package_list.keys()):
-            self.container.add_bash_command('emaint sync -r ros-overlay')
```

Comment:
This is probably just my lack of understanding, why does emaint sync need to be run for every package?

This isn't necessarily related to this PR. I'm only just looking at it now. Does each package have a different context that the sync runs inside of or could the sync happen once before iterating over all packages?

Reply:
Based on my earlier statement an additional question came up in my head.
As it seems the same container is used throughout for all tests of all packages of all distros. If this is correct, I would wonder if the dependency resolution of one build could taint the environment for the same package from another distro?
In that case, it would be worthwhile finding a way to build all packages in a fresh environment to avoid such side-effects from the beginning.